### PR TITLE
configurable ZMQ message limit

### PR DESCRIPTION
**Pull request by Alex-CodeLab**
ZMQ  pub/sub has a default limit of 1000 buffered messages. When it hits the limit messages are dropped. This happens sometimes when a new block is found and ~2500messages ( 2.5MB) are send (almost) at once.  
50-800 usually got lost on my (cheap) node. 
 [ZMQ_SNDHWM](https://github.com/zeromq/libzmq/blob/master/doc/zmq_setsockopt.txt#L840) 

The PUBlisher limit,  SNDHWM, is disabled by setting it to 0.
On very slow networks it might fill up memory a bit, but that is hardly noticeable since the queue is usually emptied in a few hundred milliseconds. 
 
Same applies to the SUBscriber (although the bottleneck there is only the local processing, not the network throughput), so it would be good practice to set RCV_HWM to zero as well for some client script, although I was only able to hit the limit on the subscriber side after adding sleep(1) on every rawtx.

I've tested for about a day and had a bunch of subscribers running on different computers, and not a single message got dropped. ( also while using swap and 100% cpu  `stress --vm-bytes $(awk '/MemAvailable/{printf "%d\n", $2 * 1.9;}' < /proc/meminfo)k --vm-keep -m 1 -c 1 -i 2` )
I forced one subscriber to lag about 10000 messages , and they were still not dropped. 

I'd appreciate it if others test their performance and load. You can just use this script and check for missing messages while bitcoind is running. (for testing max throughput, stop bitcoind for a while, then restart it; it will send lots of messages while syncing)

subscriber testscript:
https://gist.github.com/FeedTheWeb/e6243410cb463759d5b29fae24550f36

### CODE REVIEW

**Comment by promag** [comment]
Not sure about disabling the limit.

**Comment by jonasschnelli** [comment]
I think a viable long term alternative is replacing ZMQ with a stateless longpolling similar to #13262.
Something like a `list_blocks_since_block` where it idles if the newest block was requested. Seems to be much saner then hoping ZMQ messages arrive.

**Comment by FeedTheWeb** [comment]
@promag I tried a 10000 limit first, and that also works (for now). But the default of 1000 seems too low

**Comment by practicalswift** [comment]
Would it make sense to bump the limit to say 2 * "expected maximum under normal operations" instead of removing the limit?

That way we won't drop messages under normal circumstances and we still get the "avoid going berserk" safety the hard upper limit is meant to provide?

**Comment by promag** [comment]
What is the "expected maximum under normal operations"?

**Comment by practicalswift** [comment]
@promag Sorry for being vague. I meant the maximum number of buffered messages that can be expected under normal operations :-)

**Comment by promag** [comment]
@practicalswift reorgs and rescans will trigger the notifiers so I don't know if we should go that way. 

Also, from the docs:
> If this limit has been reached the socket shall enter an exceptional state and
> depending on the socket type, 0MQ shall take appropriate action such as
> blocking or dropping sent messages

IMO making this configurable is the best bet and ideally should be different per notifier.

**Comment by FeedTheWeb** [comment]
Agree, making it configurable would be best. 
But also an increased for the default value, I propose 100000 (~100MB).

**Comment by laanwj** [comment]
Ping @jmcorgan

**Comment by FeedTheWeb** [comment]
@promag, it could only be made configurable per socket (not per notifier), but tbh I don't see why that would be useful. 

@practicalswift the 'normal circumstances' is not relevant here, this is only about peaks (with a duration of usually a few ms). In normal situations there never is a queue; the queue grows only when bitcoind pushes the messages faster then your network can handle (note: 'network', not client software - the receiving zmq-socket has its own buffer).

**Comment by jmcorgan** [comment]
I agree on having a config option to set this. The reason for leaving the default in the original implementation was to ensure that there were no performance impact or other side effects on bitcoind operation from malicious or poorly executing clients.

**Comment by jmcorgan** [comment]
To elaborate--ZMQ's high watermark buffered message limit prevents the ZMQ sending process from blocking when downstream resources (network bandwidth) are insufficient to handle the message volume. (The remote end can detect this when sequence numbers get skipped.) 

This was originally chosen by design to be left at the default in order to be *conservative*--enabling ZMQ notifications and connecting a client should *not* have any potential adverse impact on bitcoind operation.

I think allowing this behavior to be tuned via a command line or config file option, to set the numerical value of the high water mark, including turning it off, would be very useful for people relying on this interface. Unfortunately, it is a per-socket behavior and not per-notification, but that's the best that can be done.

**Comment by sangaman** [comment]
Making this configurable seems like the way to go. And if it spikes around 3000 messages per new block, raising the default to slightly higher than that (maybe 5000?) might be good as well.

**Comment by reel** [comment]
I had 3 subscribers on commodity HW and they were missing packets (they were LND nodes) and desycing every 2-3 blocks. Since I applied this PR, I can easily run 7-8 clients and not missing anything. Make it configurable or disable it. 

Great PR!

**Comment by jmcorgan** [comment]
@FeedTheWeb Are you planning to update this to make this configurable?

**Comment by FeedTheWeb** [comment]
@jmcorgan , we still need to agree on setting a different default value .
(10k is reasonable, imo )

**Comment by reel** [comment]
Is there any way we can poll to see the actual usage? As I understand, this is affected by the slowest peer?

**Comment by reel** [comment]
The ZMQ api is well documented. (http://api.zeromq.org/2-1:zmq-setsockopt)

ZMQ_HWM and ZMQ_SWAP could be configurable (advanced mode).

HWM default could stay at 1000 or be increased to 10000
SWAP could still default to 0 but it could be a good safety net for delicate operations

Let me know if you want me to give it a try or collaborate on this.

**Comment by FeedTheWeb** [comment]
@reel no, that is version 2.1 we use 4.*  and it does not use SWAP anymore. http://api.zeromq.org/4-1:zmq-setsockopt

**Comment by reel** [comment]
Correct, I need a coffee fix.

Defaults were set to 1000 7 years ago...

https://github.com/zeromq/libzmq/commit/d31792e652cc9fd3bc84e2abd89f232d273b7ede

**Comment by slush0** [comment]
I confirm removing this limit fixed most of weird issues with my LND node.

File: src/init.cpp [hunk 1]
```diff
@@ -454,6 +454,7 @@ void SetupServerArgs()
     gArgs.AddArg("-zmqpubhashtx=<address>", "Enable publish hash transaction in <address>", false, OptionsCategory::ZMQ);
     gArgs.AddArg("-zmqpubrawblock=<address>", "Enable publish raw block in <address>", false, OptionsCategory::ZMQ);
     gArgs.AddArg("-zmqpubrawtx=<address>", "Enable publish raw transaction in <address>", false, OptionsCategory::ZMQ);
+    gArgs.AddArg("-zmqhwm=<n>", "Set HWM limit <n> (default: 10000)", false, OptionsCategory::ZMQ);
```
Comment (jmcorgan) [inline src/init.cpp]: This will need a hidden_args entry.

File: src/zmq/zmqpublishnotifier.cpp [hunk 2]
```diff
@@ -75,9 +75,13 @@ bool CZMQAbstractPublishNotifier::Initialize(void *pcontext)
             zmqError("Failed to create socket");
             return false;
         }
-        
-        int HWMlimit = 0;
-        zmq_setsockopt(psocket, ZMQ_SNDHWM, &HWMlimit, sizeof(HWMlimit));
+
+        int hwm = gArgs.GetArg("-zmqhwm", 10000);
+        zmq_setsockopt(psocket, ZMQ_SNDHWM, &hwm, sizeof(hwm));
```
Comment (jmcorgan) [inline src/zmq/zmqpublishnotifier.cpp]: The error return should be checked here in case an invalid value is set by the user config.

**Comment by jmcorgan** [comment]
In addition to the individual comments above, there also needs to be some changes related to adding a new config option (documentation, mostly.) I suggest you do a 'git grep zmqpub' from the top-level directory (above src) to see where this is necessary.

File: src/zmq/zmqpublishnotifier.cpp (hunk 2 above)
Comment (FeedTheWeb) [reply src/zmq/zmqpublishnotifier.cpp]: if any invalid value is given, it defaults to1000.

**Comment by FeedTheWeb** [comment]
pls improve the code if you have any suggestions.
(I'm not a C++ coder, so my knowledge kinda ends here anyway)

**Comment by jmcorgan** [comment]
This would have needed a clean rebase for merge anyway, so if you don't mind I'll just implement this from scratch in a new PR.

**Comment by FeedTheWeb** [comment]
@jmcorgan , 
it has a been a few months, and I have not seen your PR. 
As I told you also in private, I welcome any improvement. 
However, your comment:  "ensure that there were no performance impact or other side effects on bitcoind operation from malicious or poorly executing clients" tells me that you don' t really understand how ZMQ works, since (when implemented correctly) these are completely separated processes. 
Anyone with a basic knowledge of such messaging systems knows that the main goal of these  'separation of concerns'  (messaging systems) is to prevent having an impact on client/3rdparty processes (in this case, bitcoind). The performance of the client simply does not affect the server (bitcoind). 
(again, not talking about network performance here,  as pointed out earlier).

From here it seems you wanted to claim ownership over this part of the code. Which fine with me - really.- (I don't even use my own name. take all the credit if you want...).  But now it seems any progress on this part is halted by you, for unknown reasons.

so... code or it didn't happen ;)

**Comment by promag** [comment]
But now it seems any progress on this part is halted by you, for unknown reasons.

@FeedTheWeb if you need this then feel free to implement it, I'd review it too.

**Comment by jmcorgan** [comment]
@FeedTheWeb Your understanding of how to motivate others to do free work to your liking seems lacking.

You made a pull request, and when I reviewed and suggested simple changes that would be needed, you immediately gave up. 

I offered to do it, and when it wasn't done in the time frame you wanted, you then made odd claims about my level of understanding of something completely unrelated to either the feature needed or the feedback I gave you on your PR.

You're also saying that I want to "claim ownership" of this section of code. I've done nothing of the sort, of course.

The code to allowing the high water mark to be configurable is easy, and the choosing the default value is something that can be discussed here for consensus by interested parties. Again, if you want to contribute, feel free to continue what you started earlier, and be prepared to address feedback until those with merge rights deem it acceptable to introduce into the tree and maintain it thereafter.

Otherwise, your attitude is baffling and not very effective at bringing about the changes you want to see through the volunteer effort of other people.

**Comment by FeedTheWeb** [comment]
thanks for the reply, @jmcorgan 
First of all, it is a volunteer project for me as well. Tbh, I'm not even that interested in this change, but 
the LND folks have been complaining about it for over a year, and it seems important, so I investigated the issue, made a case and provided evidence. 
(Why? because I worked with ZMQ before, so I knew about this issue for some years already.  And I like to fix stuff.)

My task is to identify problems (not a popular thing to do, but someone has to do it).
What I've done is identifying a issue that was a big question-mark for some for years and provided evidence. I've mentioned the issue a few times on different platforms, but it seems these kind of things do not communicate well in a chat. As a result, I wrote it down in a ticket, with proof, and some code.

I created some code only to make a case - I don't claim any proficiency in C++  so I want others to improve it. It should be low-hanging fruit for everyone else. 

Over the last months I've patiently waited and tried to contact you a few times (as have others) but never got a response. (that is just rude)
I've created the code to make it configurable, as suggested by some - even though we all know that is pointless without figuring out FIRST what a functional HWM value should be anyway;  it doesn't even solve the issue, and no one who suggested it made a case for why it would solve anything...   

I've asked you (or anyone else) to suggest a better a default HWM value; never got a reply. The reason why seems quite obvious: no one actually tested what the right value is on their system.
The current '1000' is a arbitrary value, which doesn't make any sense when you think about it. 

(as a scientist) I ask you to defend why '1000' is the correct value

**Comment by mruddy** [comment]
@FeedTheWeb chillax bro, we're all in this together. i built on your work and, i believe, covered all of the feedback so far on my branch in one commit at https://github.com/mruddy/bitcoin/tree/zmqhwm.

@jmcorgan does my commit cover your feedback? my changes create a configurable hwm option for each notification type. i left the default at 1000 since it's configurable and people that don't configure it get the current 1000 messages for all notification types.

if my commit is good, you can use it here, or i can create a separate PR. it doesn't matter to me.

**Comment by reel** [comment]
Enough with the drama.

@mruddy can you please make a new PR referencing this one so we can comment/review?

Good work everyone!

One a side note, work done on the LND side to limit the amount of connections and broadcast internally really helped. This PR will really close this issue and prevent maintaining Bitcoin forks for many of us. Thanks again!

File: src/zmq/zmqpublishnotifier.cpp [hunk 3]
```diff
@@ -76,6 +76,13 @@ bool CZMQAbstractPublishNotifier::Initialize(void *pcontext)
             return false;
         }
 
+        int hwm = gArgs.GetArg("-zmqhwm", 10000);
```
Comment (practicalswift) [inline src/zmq/zmqpublishnotifier.cpp]: Make this explicit conversion explicit :-)

**Comment by laanwj** [comment]
Closed by #14060, I guess.

**Comment by mruddy** [comment]
yes, thanks!

**Pull request closed without merging**